File: python/211.py
class WordDictionary(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.root = {}

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        p = self.root
        for c in word:
            if c not in p:
                p[c] = {}
            p = p[c]
        p[True] = True

    def search(self, word, root=None):
        """
        Returns if the word is in the data structure. A word could
        contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        p = self.root if root is None else root
        for i, c in enumerate(word):
            if c == '.':
                for pc in p.values():
                    if pc != True and self.search(word[i+1:], pc):
                        return True
                return False
            else:
                if c not in p:
                    return False
                p = p[c]
        return True in p

# Your WordDictionary object will be instantiated and called as such:
wordDictionary = WordDictionary()
wordDictionary.addWord("word")
print(wordDictionary.search("word"))
print(wordDictionary.search("w.rd"))
print(wordDictionary.search("pattern"))

File: cpp/211.cpp
#include <string>
using namespace std;

class Node {
public:
  bool end;
  Node **ch;
  Node() {
    ch = new Node *[26]();
    end = false;
  }
};

class WordDictionary {
  Node *root;

public:
  WordDictionary() { root = new Node(); }

  // Adds a word into the data structure.
  void addWord(string word) {
    Node *p = root;
    for (char c : word) {
      int id = c - 'a';
      if (p->ch[id] == nullptr)
        p->ch[id] = new Node();
      p = p->ch[id];
    }
    p->end = true;
  }

  // Returns if the word is in the data structure. A word could
  // contain the dot character '.' to represent any one letter.
  bool search(string word) { return search(word, 0, root); }

  bool search(string word, size_t i, Node *p) {
    for (; i < word.length(); i++) {
      if (word[i] == '.') {
        for (int j = 0; j < 26; j++)
          if (p->ch[j] != nullptr && search(word, i + 1, p->ch[j]))
            return true;
        return false;
      } else {
        int id = word[i] - 'a';
        if (p->ch[id] == nullptr)
          return false;
        else
          p = p->ch[id];
      }
    }
    return p && p->end;
  }
};

// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");
